### DISCUSSION TITLE
How to use the hardware SPI On ESP32-S3 in Arduino IDE

### DISCUSSION THREAD

**Discussion by LiwareTech** [olikraus/u8g2#2457] [Q&A] [answered]
I have purchased a monochrome display screen driven by the SSD1322 chip. It has a resolution of 256*64 and a 3-wire SPI interface. I bought it on Alibaba (a Chinese online store) from this link: "https://detail.1688.com/offer/520064519036.html?spm=a2615.2177701.autotrace-offerGeneral.20.2ab74cdaEPnmfk". I want to use it to create a calculator. However, I am experiencing an issue with the slow refresh rate of the screen. It takes approximately 700-1000 milliseconds to refresh once. I am using the Arduino IDE and have tried various solutions found online, but none of them have been successful. Could someone please advise me on how to improve the refresh rate of the screen? Thank you very much.(translated by ChatGPT)

**Comment by olikraus** [comment, marked as answer]
Which u8g2 constructor did you use? 
Ensure to use a HW SPI contructor!

**Reply by LiwareTech** [reply]
Hello, I tried using the hardware SPI constructor, but it didn't work after uploading. The constructor looks like this: `U8G2_SSD1322_NHD_256X64_F_4W_HW_SPI u8g2(U8G2_R0, /* cs=*/ 10, /* dc=*/ 9, /* reset=*/ 8);`. In this constructor, the corresponding MOSI, MISO, CLK pins are not specified. The ESP32-S3 has four sets of SPI, with two hardware SPI used internally in the module and only one hardware SPI brought out. The OLED connected to this SPI does not display anything. I suspect that U8g2 might be using the internal SPI of the module, causing my OLED screen to not receive any signal.
Can you help me?Thank you.

**Reply by olikraus** [reply]
> MOSI, MISO, CLK pins are not specified

Yes, because u8g2 will use exactly the same pins as used by the Arduino IDE/Framework SPI.h lib.

**Reply by LiwareTech** [reply]
So what can I do?

**Reply by olikraus** [reply]
Connect your displays to those pins, which are used and supported by the Arduino SPI object. The pins are listed here: https://www.arduino.cc/reference/en/language/functions/communication/spi/ 
If your board is missing, then you need to check your board related documentation.

**Reply by LiwareTech** [reply]
Thank you very much. Some people prefer using small System on Chips (SoCs), such as the ESP32 series. They have SPI pins that can be mapped customarily. For example, this one: <https://www.espressif.com.cn/sites/default/files/documentation/esp32-s3-wroom-1_wroom-1u_datasheet_en.pdf>, which comes from Espressif Systems. Could you add some mapping information to the HW constructor, like so: `U8G2_SSD1322_NHD_256X64_F_4W_HW_SPI u8g2(U8G2_R0, /* MOSI */, /* MISO */, /* SCK */, /* cs=*/ 10, /* dc=*/ 9, /* reset=*/ 8);`? Similar to how the TFT_eSPI library does it. If you can add this line, I would greatly appreciate it. Love from China.

**Reply by olikraus** [reply]
such a mapping exists already, see the notes below the table here: https://github.com/olikraus/u8g2/wiki/u8g2setupcpp#communication
I think the same mapping is also possible with SPI.

**Reply by LiwareTech** [reply]
Thank you very much.

I have something to share you.
On bilibili(a Chinese video webside),SHA256 make a calculator named "SCI_CALC".(https://www.bilibili.com/video/BV1eB4y1d7L4/?spm_id_from=333.788&vd_source=f4c9322e9f2f8f4080f2b9b147d60dd7.
and Github address:https://github.com/shaoxiongduan/sci-calc)
It's so great.

**Reply by olikraus** [reply]
Nice, also some games are included :-D

**Comment by LiwareTech** [comment]
Hello, I have found where the problem lies and now I have resolved it. I want to put the solution here, in case someone needs it. The hardware SPI pins I found online for ESP32-S3 are GPIO11, GPIO12, and GPIO13. However, the default SPI pins for the "ESP32-S3 Dev Module" downloaded from Arduino IDE are not these. Due to the IO MUX of ESP32-S3, SPI pins can be freely allocated. So all you need to do is open C: \ Users \ "Your user name" \ AppData \ Local \ Arduino 15 \ packages \ esp32 \ hardware \ esp32\ **version number** \ variables \ adafruit_feather_ esp32s3 \ pins_arduino. h and modify it:
﻿
static const uint8_t MOSI  = 11;
static const uint8_t SCK   = 12;
static const uint8_t MISO  = 13;
﻿
These few items are sufficient. Then you can wire according to the corresponding GPIO.
﻿
Thank you for providing the solution
Lemon Studio

**Reply by olikraus** [reply]
Thanks for the information!
